### PR TITLE
feat: add a plugin to integrate with Good Ghosting

### LINKED ISSUE #3337: [Bounty] Integrate GoodGhosting, a savings game, in Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 400 $MASK or $2000

## Part 1: Problem Description

**Background**
Goodghosting is a DeFi pool which rewards regular saving. It's an social savings game that active participants will reap interest from those who fail to put down savings in time. We think it has really good synergy with traditional social networks and thus worth the integration. 

Previously defi plugins on Mask, like the trading widgets powered by 0x and Uniswap, has been plagued by high ETH gas prices.  However with the launch of polygon and BSC integration, the problem should have been largely resolved. Savings app which was hindered by monstrous ETH gas price is now very much affordable. This could be Mask's first Polygon Defi plugin.

Goodghosting has received grants from AAVE and the Graph. It participated in Gitcoin grants and received donations from Andre Cronje and Defi Dad and more than 100 others 

**Solution**
Create a plugin that integrate Good Ghosting. It could look like the Gitcoin donation widget/card with a few tabs displaying different info

image
_The Gitcoin widget/card_

image
_Example dashboard from Goodghosting_

image
_Example personal stats page_

One tab could be game stats like shown above, second one personal stats, third one timeline, fourth one other players.

The tab order can be changed, corresponding landing and ending page needs to be added. Share button for such a social game is a must. Some simplification is probably needed compared to the website view.

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://goodghosting.com/#/
- https://docs.goodghosting.com/goodghosting/guides/connecting#connecting-to-goodghosting-on-polygon
- https://gitcoin.co/grants/1112/goodghosting-a-defi-savings-game

## Part 3: Contact and Support
@yisiliu for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q
Goodghosting discord server: https://discord.gg/qpWKgxsABJ

**Submission Requirements**
Please include demo to show the integration is working.

**Winner Announcement Date**
Winners will be announced at the conclusion of Gitcoin 10 Hackathon

**Judging Criteria**
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by OsamaElHariri**
closes #3337

Add a plugin to integrate with [GoodGhosting](https://goodghosting.com/#/), a DeFi savings game dApp on Polygon

All data is retrieved from the smart contracts, similarly to how the data is currently retrieved on the GoodGhosting web page

Once this plugin is enabled, you can see it in action on this [twitter post](https://twitter.com/goodghosting/status/1405204093254221824)

### Video Demo

https://user-images.githubusercontent.com/23722804/124515484-3a961e80-dde8-11eb-82c8-b9e00d199775.mp4

---
### Added features after testing with GoodGhosting (more info in [this comment](https://github.com/DimensionDev/Maskbook/pull/3570#issuecomment-886037411) ):
- Allow players to join the game, make deposits, and withdraw their earnings.
- Allow players to withdraw early from the game before it ends
- Get the GoodGhosting contract address dynamically by fetching this config file: https://github.com/Good-Ghosting/mask-plugin-integration/blob/master/games.json

### CODE REVIEW

**Comment by septs** [comment]
plz rename this pr name, thankd

File: packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:77]:
```suggestion
                            title={playerInfo.winning}
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:85]:
```suggestion
                            title={playerInfo.waiting}
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:90]:
```suggestion
            <Grid className={classes.infoRow} container justifyContent="center">
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:95]:
```suggestion
                            title={playerInfo.ghosts}
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:103]:
```suggestion
                            title={playerInfo.dropouts}
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:57]: extract as function

File: packages/maskbook/src/plugins/GoodGhosting/UI/TimelineView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/TimelineView.tsx:174]:
```suggestion
                                    )} />
```

File: packages/maskbook/src/plugins/GoodGhosting/hooks/useGameInfo.ts
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/hooks/useGameInfo.ts:65]:
```suggestion
            firstSegmentStart: Number.parseInt(firstSegmentStart, 10),
            currentSegment: Number.parseInt(currentSegment, 10),
            lastSegment: Number.parseInt(lastSegment, 10),
            segmentLength: Number.parseInt(segmentLength, 10),
            numberOfPlayers: Number.parseInt(numberOfPlayers, 10),
```

File: packages/maskbook/src/plugins/GoodGhosting/utils.ts
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/utils.ts:17]:
```suggestion
    if (mostRecentSegmentPaid < currentSegment - 1) return PlayerStatus.Ghost
    if (mostRecentSegmentPaid === currentSegment - 1) return PlayerStatus.Waiting
    if (mostRecentSegmentPaid === currentSegment) return PlayerStatus.Winning
```

**Comment by OsamaElHariri** [comment]
Updated the PR based on the suggestions.

Also, I added a timer on top of the tabs similar to what shows on the GoodGhosting website.

File: packages/maskbook/src/plugins/GoodGhosting/README.md
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/README.md:1]:
```suggestion
# Plugin: Good Ghosting
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:55]: plz move this function to this file bottom, thanks

File: packages/maskbook/src/plugins/PluginUI.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/PluginUI.ts:16]: Hello, @OsamaElHariri thanks for your contribution. Please use the new way to define a plugin, you can see many examples in the plugins folder. `dhedge` is the only last plugin not migrated to the new definition standard and we should not add a new definition here.

Comment (Jack-Works) [reply packages/maskbook/src/plugins/PluginUI.ts:16]: FYI, all new APIs are well documented in packages\plugin-infra\src\types.ts and the new plugin structure can be seen in packages\plugins\example.

Comment (OsamaElHariri) [reply packages/maskbook/src/plugins/PluginUI.ts]: I went over the example of how the updated plugin infra looks like and updated the PR. This plugin is now defined using the updated way of handling plugins

**Review by Jack-Works — APPROVED** [review]
Need test from @AinuX and blockchain review from @zhouhanseng or @guanbinrui

File: packages/maskbook/src/plugins/GoodGhosting/utils.ts
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/utils.ts:33]: why function name not is `isEndOfTimeline`

**Comment by AinuX** [comment]
Just checked “Game”, “Timeline” and “Everyone” tab under Polygon and other unsupported chains, seems OK to me.

But, since there is only one active game pool and have already started, so I can’t to join the game to check the “Personal” tab, and also have no more game pool to test.

One concern, the plugin will be rendered when any url prefixed with “https://goodghosting.com”, even like url “https://goodghosting.com/#/2#faqs”.

File: packages/maskbook/src/plugins/GoodGhosting/utils.ts
Comment (OsamaElHariri) [reply packages/maskbook/src/plugins/GoodGhosting/utils.ts]: Updated the function name :+1:

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/plugins/GoodGhosting/UI/PreviewCard.tsx
Comment (guanbinrui) [inline packages/maskbook/src/plugins/GoodGhosting/UI/PreviewCard.tsx:54]:
```suggestion
    if (loading) {
        return <Typography color="textPrimary">Loading...</Typography>
    } 
    
    if (error || !info) {
        return (
            <Box display="flex" flexDirection="column" alignItems="center">
                <Typography color="textPrimary">Something went wrong.</Typography>
                <Button sx={{ marginTop: 1 }} size="small" onClick={retry}>
                    Retry
                </Button>
            </Box>
        )
```

**Review by guanbinrui — APPROVED** [review]

**Comment by Jack-Works** [comment]
hi here some merge conflicts

File: packages/maskbook/src/_locales/en/messages.json
Comment (guanbinrui) [inline packages/maskbook/src/_locales/en/messages.json:787]:
```suggestion
    "plugin_good_ghosting_not_a_participant": "Looks like you're not a participant in this game。",
```

File: packages/maskbook/src/_locales/en/messages.json
Comment (guanbinrui) [inline packages/maskbook/src/_locales/en/messages.json:789]:
```suggestion
    "plugin_good_ghosting_game_end": "This game has ended.",
```

File: packages/maskbook/src/_locales/en/messages.json
Comment (guanbinrui) [inline packages/maskbook/src/_locales/en/messages.json:795]:
```suggestion
    "plugin_good_ghosting_deposit_help_text": "Deposit {{segmentPayment}} {{token}} for this round and continue to be a winner.",
```

**Comment by OsamaElHariri** [comment]
After testing with GoodGhosting and discussing the use case of this plugin, it became clear that a couple of more features are needed. The features added are the following (I'll also add these to the PR description above):
- Allow players to join the game, make deposits, and withdraw their earnings.
- Allow players to withdraw early from the game before it ends
- Get the GoodGhosting contract address dynamically by fetching this config file: https://github.com/Good-Ghosting/mask-plugin-integration/blob/master/games.json

Here are some images that show the new additions in the UI:

**Joining a game (make deposit and withdraw actions look similair to this):**
image
image


**Withdraw early:**
image

File: packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:72]:
```suggestion
    onUpdateTargetDate: () => void
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx:3]: plz split to multi import statement

CC @Jack-Works

File: packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx:57]: move this function to this file bottom

File: packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:57]: plz note why need ignore error

Comment (septs) [reply packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:57]:
```suggestion
        } catch (error) {
            // ignore error
            // HACK: ....
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:139]:
```suggestion
function usePlayerStatusMessage(info: GoodGhostingInfo, player?: Player) {
```

File: packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:2]: plz split to multi import statement

Comment (septs) [reply packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:2]:
```suggestion
import intervalToDuration from 'date-fns/intervalToDuration'
import isBefore from 'date-fns/isBefore'
```

**Review by septs — CHANGES_REQUESTED** [review]
plz resolve all warnings

File: packages/maskbook/src/_locales/en/messages.json
Comment (OsamaElHariri) [reply packages/maskbook/src/_locales/en/messages.json:779]: Will keep this one as is since it's a title

---

image

File: packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx
Comment (OsamaElHariri) [reply packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx]: Split the date-fns imports to individual statements :+1:

**Comment by OsamaElHariri** [comment]
Thanks for the review @septs 

I split the date imports into individual statements, fixed the function names and locations, handled the error instead of ignoring, and fixed the pipeline linter warnings.

File: packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:5]: ?

**Review by guanbinrui — APPROVED** [review]

**Review by Jack-Works — APPROVED** [review]

File: packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:4]:
```suggestion
import formatDuration from 'date-fns/formatDuration'
```

File: packages/maskbook/src/plugins/GoodGhosting/types.tsx
Comment (septs) [inline packages/maskbook/src/plugins/GoodGhosting/types.tsx:27]:
```suggestion
export interface Player {
    addr: string
    amountPaid: string
    canRejoin: boolean
    mostRecentSegmentPaid: string
    withdrawn: boolean
}
```

**Comment by Jack-Works** [comment]
Hi please merge with develop branch, I have moved `createPluginMessage` and `createPluginRPC` into the `@masknet/plugin-infra` package.

**Comment by OsamaElHariri** [comment]
I applied the changes and rebased with develop

**Review by septs — APPROVED** [review]
approve, but plz wait for @Jack-Works and @guanbinrui re-check all files

**Review by guanbinrui — APPROVED** [review]
LGTM

**Review by Jack-Works — APPROVED** [review]

**Comment by septs** [comment]
@OsamaElHariri plz wait for @guanbinrui merge this branch to develop

**Pull request merged**
